compiler/src/iree/compiler/Dialect/Flow/TransformExtensions/FlowExtensions.cpp

```diff
@@ -693,6 +693,118 @@ transform_dialect::ClonePrecedingOpIntoDispatchRegionOp::apply(
 }
 
+// Clone a `target` op that is succeeding the given dispatch region op into the
+// dispatch region.
+//
+// All operands of the target are replaced with values defined inside of the
+// dispatch region when possible.
+//
+// If `updateUsesOutsideOfRegion` is set, all uses of the target op after the
+// dispatch region, are updated: The target op's results are returned from the
+// the dispatch region an used in those places.
+//
+// Example when `updateUsesOutsideOfRegion` is set:
+//
+// %r = flow.dispatch.region -> (tensor<?xf32>{%d0}) {
+//   %1 = "another_op"() : () -> (tensor<?xf32>)
+//   flow.return %1 : tensor<?xf32>
+// }
+// %0 = "some_op"(%r) : (tensor<?xf32>) -> (tensor<?xf32>)
+// %2 = "yet_another_use"(%0) : (tensor<?xf32>) -> (tensor<?xf32>)
+//
+// In this example, "some_op" will be cloned into the dispatch region and the
+// OpOperand of "yet_another_use" will be replaced:
+//
+// %r:2 = flow.dispatch.region -> (tensor<?xf32>{%d0}) {
+//   %1 = "another_op"() : () -> (tensor<?xf32>)
+//   %0_clone = "some_op"(%1) : (tensor<?xf32>) -> (tensor<?xf32>)
+//   flow.return %1, %0_clone : tensor<?xf32>, tensor<?xf32>
+// }
+// %2 = "yet_another_use"(%r#1) : (tensor<?xf32>) -> (tensor<?xf32>)
+static FailureOr<Flow::DispatchRegionOp> cloneSucceedingOpIntoDispatchRegion(
+    RewriterBase &rewriter, Operation *target, Flow::DispatchRegionOp regionOp,
+    bool updateUsesOutsideOfRegion) {
+  assert(regionOp->isBeforeInBlock(target) &&
+         "expected that region op comes first");
+  Block &body = regionOp.getBody().front();
+
+  // Gather all uses of `target`.
+  SmallVector<OpOperand *> usesOutsideOfRegion;
+  for (OpOperand &use : target->getUses()) usesOutsideOfRegion.push_back(&use);
+
+  // Clone op into dispatch region.
+  OpBuilder::InsertionGuard guard(rewriter);
+  auto returnOp = cast<Flow::ReturnOp>(body.getTerminator());
+  rewriter.setInsertionPoint(returnOp);
+  Operation *newTargetOp = rewriter.clone(*target);
+
+  // Replace all operands that are results of the regionOp.
+  for (OpOperand &operand : newTargetOp->getOpOperands()) {
+    if (operand.get().getDefiningOp() == regionOp) {
+      unsigned resultNumber = operand.get().cast<OpResult>().getResultNumber();
+      operand.set(returnOp->getOperand(resultNumber));
+    }
+  }
+
+  // Replace all uses outside of the dispatch region.
+  if (updateUsesOutsideOfRegion) {
+    unsigned previousNumResults = regionOp->getNumResults();
+
+    // Note: Appending results one-by-one here so that this can be extended to
+    // specific results in the future. Many ops have just one result, so this
+    // should not be a large overhead.
+    for (Value v : newTargetOp->getResults()) {
+      auto newRegionOp = appendDispatchRegionResult(rewriter, regionOp, v);
+      if (failed(newRegionOp)) return failure();
+      regionOp = *newRegionOp;
+    }
+
+    // Replace uses of `target` after the dispatch region.
+    for (OpOperand *use : usesOutsideOfRegion) {
```

Comment:
I dont see much value in pre-caching the uses this way? You might as well just iterate over the uses right here?

Reply:
It this safe? I ran into issues when modifying uses like that in the past. (Probably because uses are going to disappear while I'm iterating over all uses.)

Reply:
You dont need to pre-cache uses all the way at the start. You can just get the uses before iterating over it. Lot of the logic here will get simplified.